finality-aleph/src/testing/data_io.rs

```diff
@@ -0,0 +1,385 @@
+
+    // Build 4 new blocks different from finalized
+    let mut blocks = Vec::new();
+    for i in 0u64..4 {
+        let mut digest = Digest::default();
+        digest.push(sp_runtime::generic::DigestItem::Other(
+            i.to_le_bytes().to_vec(),
+        ));
+        let block = client
+            .new_block_at(&BlockId::Number(i), digest, false)
+            .unwrap()
+            .build()
+            .unwrap()
+            .block;
+        blocks.push(AlephData::new(block.header.hash(), block.header.number));
+    }
+
+    // Send the created blocks
+    network_tx.unbounded_send(blocks.clone()).unwrap();
+
+    // Blocks should be sent to network
+    let message = network.next().await.unwrap();
+    assert_eq!(message.included_blocks(), blocks);
+
+    exit_data_store_tx.send(()).unwrap();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn does_not_send_messages_without_import() {
+    let (
+        task_handle,
+        mut client,
+        DataStoreChannels {
+            mut network,
+            block_requests_rx: _,
+            network_tx,
+            exit_data_store_tx,
+        },
+    ) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+
+    // Import 4 blocks
+    let blocks = import_blocks(&mut client, 4, false).await;
+
+    // Build new block
+    let not_imported_block = client
+        .new_block(Default::default())
+        .unwrap()
+        .build()
+        .unwrap()
+        .block;
+
+    // Send new block
+    network_tx
+        .unbounded_send(vec![AlephData::new(
+            not_imported_block.header.hash(),
+            not_imported_block.header.number,
+        )])
+        .unwrap();
+
+    // Send imported blocks
+    network_tx.unbounded_send(blocks.clone()).unwrap();
+
+    // Should receive imported blocks
+    let message = network.next().await.unwrap();
+    assert_eq!(message.included_blocks(), blocks);
+
+    // Close Data Store
+    exit_data_store_tx.send(()).unwrap();
+
+    // Network should have no other block than already imported
+    let message = network.next().await;
+    assert!(message.is_none());
+
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn sends_block_request_on_missing_block() {
+    let (
+        task_handle,
+        client,
+        DataStoreChannels {
+            network: _,
+            mut block_requests_rx,
+            network_tx,
+            exit_data_store_tx,
+        },
+    ) = prepare_data_store();
+
+    let data_store_handle = tokio::spawn(task_handle);
+
+    // Build new block
+    let block = client
+        .new_block(Default::default())
+        .unwrap()
+        .build()
+        .unwrap()
+        .block;
+    let data = AlephData::new(block.header.hash(), block.header.number);
+
+    // Send new block
+    network_tx.unbounded_send(vec![data]).unwrap();
+
+    // New block should be requested
+    let requested_block = block_requests_rx.next().await.unwrap();
```

Comment:
Nit: I'm comparing previous version to this one, and IMHO `expect` is more readable - this clearly states intention to a reader and on top of that in the possible scenario that this test fails it states it in stacktrace. 
```rust
 let requested_block = block_requests_rx
        .next()
        .await
        .expect("Block should be requested");
```

Reply:
Well, the thing is that when the block is not requested there will be no error and the test will just run forever. The only case when this except prints something is when all sending endpoints are dropped. Which is very unlikely to happen. That's why I removed it.

Reply:
Right, I just mean that this snippet:
```rust
// New block should be requested 
let requested_block = block_requests_rx.next().await.unwrap();
```
or this one-liner
```rust
 let requested_block = block_requests_rx.next().await.expect("Block should be requested");
```

I would say the bottom is more readable to a reader, without even knowing details of code around, since we state our intent explicitly what an exceptional scenario is (and we don't need comment now).

Reply:
Not that there is timeout there is a panic in case receive fails.